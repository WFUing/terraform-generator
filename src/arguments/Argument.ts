export default class Argument {

  readonly argument: string | Argument;
  readonly asIs: boolean;

  /**
   * Construct argument.
   * 
   * @param arg argument as string or copy from another argument object
   * @param asIs argument will be printed as is, without extra symbol, quotes and whatnot, regardless of Terraform version, default = false.
   */
  constructor(arg: string | Argument, asIs = false) {
    if (!arg || (typeof arg === 'string' && !arg.trim())) {
      throw new Error('Argument cannot be empty.');
    }

    this.argument = typeof arg === 'string' ? arg.trim() : arg;
    this.asIs = asIs;
  }

  /**
   * To Terraform representation.
   */
  toTerraform(): string {
    let str = '';
    if (this.argument instanceof Argument) {
      str += this.argument.toTerraform();
    } else {
      str += this.argument;
    }
    return str;
  }

  /**
   * To string, used when argument is used in string interpolation.
   */
  toString(): string {
    return `\${${this.toTerraform()}}`;
  }

}

/**
 * Convenient function to construct new argument.
 * 
 * @param arg argument as string or copy from another argument object
 * @param asIs argument will be printed as is, without extra symbol, quotes and whatnot, regardless of Terraform version, default = false.
 */
export const arg = (arg: string | Argument, asIs = false): Argument => new Argument(arg, asIs);
