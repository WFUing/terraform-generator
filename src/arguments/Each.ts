import { Argument } from '..';

export default class Each extends Argument {

  /**
   * Construct each argument.
   * 
   * @param attrName attribute name
   */
  constructor(attrName: string) {
    super(`each.${attrName}`);
  }

}

/**
 * Convenient function to construct new each argument.
 * 
 * @param attrName attribute name
 */
export const each = (attrName: string): Each => new Each(attrName);
