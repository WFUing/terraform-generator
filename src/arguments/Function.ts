import { Argument } from '..';
import TerraformGeneratorUtils from '../TerraformGeneratorUtils';

export default class Fn extends Argument {

  /**
   * Construct function argument.
   * 
   * @param name function name
   * @param args function arguments
   */
  constructor(name: string, ...args: any[]) {
    super(Fn.constructArgument(name, ...args));
  }

  private static constructArgument(fn: string, ...args: any[]): string {
    if (!fn || !fn.trim()) {
      throw new Error('Function name cannot be empty.');
    }
    if (args.filter(arg => arg == null).length > 0) {
      throw new Error(`Invalid function argument: ${args}`);
    }

    let str = `${fn}(`;
    args.forEach((arg, i) => {
      str += TerraformGeneratorUtils.argumentValueToString(null, arg);
      if (i < args.length - 1) {
        str += ', ';
      }
    });
    str += ')';
    return str;
  }

}

/**
 * Convenient function to construct new function.
 * 
 * @param name function name
 * @param args function arguments
 */
export const fn = (name: string, ...args: any[]): Fn => new Fn(name, ...args);
