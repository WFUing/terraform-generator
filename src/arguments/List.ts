import { Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class List extends Argument {

  /**
   * Construct list.
   *
   * @param elements list elements
   */
  constructor(...elements: any[]) {
    super(List.#constructArgument(elements));
  }

  static #constructArgument(elements: any[]): string {
    let str = '[\n';
    elements.forEach((element, i) => {
      str += `${Util.argumentValueToString(element)}${i < elements.length - 1 ? ',' : ''}\n`;
    });
    str += ']';
    return str;
  }

}

/**
 * Convenient function to construct new [[List]].
 *
 * @param elements list elements
 *
 * @category Argument
 */
export const list = (...elements: any[]): List => new List(...elements);
