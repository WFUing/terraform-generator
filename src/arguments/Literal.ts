import { Argument } from '..';

export default class Literal extends Argument {

  /**
   * Construct literal argument.
   * 
   * @param literal literal string
   */
  constructor(literal: string) {
    super(literal, true);
  }

}

/**
 * Convenient function to construct new literal argument.
 * 
 * @param literal literal string
 */
export const literal = (literal: string): Literal => new Literal(literal);
