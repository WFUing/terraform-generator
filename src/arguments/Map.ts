import { TerraformArgs, Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Map extends Argument {

  readonly arguments: TerraformArgs;

  /**
   * Construct map.
   *
   * @param args map values
   */
  constructor(args: TerraformArgs) {
    super(Map.#constructArgument(args));
    this.arguments = args;
  }

  static #constructArgument(args: TerraformArgs): string {
    return Util.argumentValueToString(args)!;
  }

}

/**
 * Convenient function to construct new [[Map]].
 *
 * @param args map values
 *
 * @category Argument
 */
export const map = (args: TerraformArgs): Map => new Map(args);
