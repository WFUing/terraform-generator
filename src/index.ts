export { BlockArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, ProvisionerArgs, ProvisionerType,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc
} from './arguments';

export {
  List, list,
  Map, map
} from './types';
