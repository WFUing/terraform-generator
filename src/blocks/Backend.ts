import { Argument, Attribute } from '../arguments';
import { TerraformArgs, Util } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export class Backend extends Block {

  readonly type: string;

  /**
   * Construct backend.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  constructor(type: string, args: TerraformArgs) {
    super('backend', [type], args, undefined, true);

    this.type = type;
  }

  override asArgument(): Argument {
    return new Argument(JSON.stringify(this.type));
  }

  override attr(_name: string): Attribute {
    throw Util.inaccessibleMethod();
  }

}
