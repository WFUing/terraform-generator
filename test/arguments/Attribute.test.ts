import { resource } from '..';
import { Attribute, attr } from '../../src/arguments';
import { Block } from '../../src/blocks';

test('Attribute invalid args', () => {
  expect(() => new Attribute(null as unknown as Block, null as unknown as string)).toThrow();
  expect(() => new Attribute(resource, null as unknown as string)).toThrow();
  expect(() => new Attribute(null as unknown as Block, 'x')).toThrow();
  expect(() => new Attribute(null as unknown as Block, '')).toThrow();
  expect(() => new Attribute(null as unknown as Block, ' ')).toThrow();
});

test('Attribute', () => {
  expect(new Attribute(resource, 'x').toTerraform()).toMatchSnapshot();
});

test('attr', () => {
  expect(attr(resource, 'x').toTerraform()).toMatchSnapshot();
});
