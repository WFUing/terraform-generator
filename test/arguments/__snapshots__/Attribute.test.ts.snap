// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attribute 1`] = `"type.name.x"`;

exports[`attr 1`] = `"type.name.x"`;
